dbos/_core/system_database.py

```diff
@@ -0,0 +1,1734 @@
+                )
+                .returning(SystemSchema.notifications.c.message)
+            )
+            rows = (c.execute(delete_stmt)).fetchall()
+            message: Any = None
+            if len(rows) > 0:
+                message = serialization.deserialize(rows[0][0])
+            self.record_operation_result_sync(
+                {
+                    "workflow_uuid": workflow_uuid,
+                    "function_id": function_id,
+                    "output": serialization.serialize(
+                        message
+                    ),  # None will be serialized to 'null'
+                    "error": None,
+                },
+                conn=c,
+            )
+        return message
+
+    async def recv_async(
+        self,
+        workflow_uuid: str,
+        function_id: int,
+        timeout_function_id: int,
+        topic: Optional[str],
+        timeout_seconds: float = 60,
+    ) -> Any:
+        topic = topic if topic is not None else _dbos_null_topic
+
+        # First, check for previous executions.
+        recorded_output = await self.check_operation_execution_async(
+            workflow_uuid, function_id
+        )
+        if recorded_output is not None:
+            dbos_logger.debug(f"Replaying recv, id: {function_id}, topic: {topic}")
+            if recorded_output["output"] is not None:
+                return serialization.deserialize(recorded_output["output"])
+            else:
+                raise Exception("No output recorded in the last recv")
+        else:
+            dbos_logger.debug(f"Running recv, id: {function_id}, topic: {topic}")
+
+        # Insert a condition to the notifications map, so the listener can notify it when a message is received.
+        payload = f"{workflow_uuid}::{topic}"
+        condition = threading.Condition()
+        # Must acquire first before adding to the map. Otherwise, the notification listener may notify it before the condition is acquired and waited.
+        condition.acquire()
+        self.notifications_map[payload] = condition
+
+        # Check if the key is already in the database. If not, wait for the notification.
+        init_recv: Sequence[Any]
+        async with self.async_engine.begin() as c:
+            init_recv = (
+                await c.execute(
+                    sa.select(
+                        SystemSchema.notifications.c.topic,
+                    ).where(
+                        SystemSchema.notifications.c.destination_uuid == workflow_uuid,
+                        SystemSchema.notifications.c.topic == topic,
+                    )
+                )
+            ).fetchall()
+
+        if len(init_recv) == 0:
+            # Wait for the notification
+            # Support OAOO sleep
+            actual_timeout = await self.sleep_async(
+                workflow_uuid, timeout_function_id, timeout_seconds, skip_sleep=True
+            )
+            condition.wait(timeout=actual_timeout)
```

Comment:
So... this is not good if the code that is supposed to do the notification is in the same event loop, because it's a deadlock then.  Probably it is not.  But it is fiddly to reason about.